toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,409 @@
+				logger.Log.Warnf("Pre-caching: '%s' failed, not found in any repos", result.pkgName)
+				unavailable++
+			}
+		case <-doneChannel:
+			// All workers are done, finish this iteration of the loop and then return
+			done = true
+		}
+
+		// Calculate the progress percentage and update the progress counter if needed (update every 'progressIncrement' percent)
+		completed := downloaded + skipped + failed + unavailable
+		progressPercent := (float64(completed) / float64(total)) * 100
+		if progressPercent > lastProgressUpdate+progressIncrement || done {
+			logger.Log.Infof("Pre-caching: %3d%% ( downloaded: %4d, skipped: %4d, unavailable: %4d, failed: %4d )", int(progressPercent), downloaded, skipped, unavailable, failed)
+			lastProgressUpdate = progressPercent
+		}
+	}
+	return
+}
+
+// precachePackage will attempt to download the specified package. It will return a downloadResult struct via the results.
+// This function runs with best effort, so it will return a result (of type downloadResultTypeFailure) if any error occurs
+// rather than returning an error. The caller is expected to have added to the provided wait group, while this function is
+// responsible for removing itself from the wait group. As much processing as possible is done before acquiring the
+// network operations semaphore to minimize the time spent holding it.
+func precachePackage(pkg *repocloner.RepoPackage, availablePackages map[string]string, outDir string, wg *sync.WaitGroup, results chan<- downloadResult, netOpsSemaphore chan struct{}) {
+	const (
+		downloadRetryAttempts = 2
+		downloadRetryDuration = time.Second
+	)
+
+	defer wg.Done()
+
+	// File names are of the form "<name>-<version>.<distro>.<arch>.rpm"
+	pkgName := fmt.Sprintf("%s-%s.%s.%s", pkg.Name, pkg.Version, pkg.Distribution, pkg.Architecture)
+	fileName := fmt.Sprintf("%s.rpm", pkgName)
+	fullFilePath := path.Join(outDir, fileName)
+	result := downloadResult{
+		pkgName: fileName,
+	}
+
+	// Bail out early if the file already exists
+	if exists, _ := file.PathExists(fullFilePath); exists {
+		result.resultType = downloadResultTypeSkipped
+		results <- result
+		return
+	}
+
+	// Get the url for the package, or bail out if it is not available
+	url, ok := availablePackages[pkgName]
+	if !ok {
+		result.resultType = donwloadResultTypeUnavailable
+		results <- result
+		return
+	}
+
+	// Limit the number of concurrent network operations to avoid overloading the network. All work past this point
+	// is either network related, or trivial, so we can safely hold the semaphore until we are done.
+	netOpsSemaphore <- struct{}{}
+	defer func() {
+		<-netOpsSemaphore
+	}()
+
+	logger.Log.Debugf("Pre-caching '%s' from '%s'", fileName, url)
+	err := retry.Run(func() error {
+		err := network.DownloadFile(url, fullFilePath, nil, nil)
+		if err != nil {
+			logger.Log.Warnf("Pre-caching: Attempt to download (%s) failed. Error: %s", url, err)
+		}
+		return err
+	}, downloadRetryAttempts, downloadRetryDuration)
```

Comment:
While trivial, why not make hold the semaphore even closer to the download logic? Assuming it's not any additional effort. One idea:
```suggestion
	logger.Log.Debugf("Pre-caching '%s' from '%s'", fileName, url)
	err := retry.Run(func() error {
		// Limit the number of concurrent network operations to avoid overloading the network.
		netOpsSemaphore <- struct{}{}
		defer func() {
			<-netOpsSemaphore
		}()

		err := network.DownloadFile(url, fullFilePath, nil, nil)
		if err != nil {
			logger.Log.Warnf("Pre-caching: Attempt to download (%s) failed. Error: %s", url, err)
		}
		return err
	}, downloadRetryAttempts, downloadRetryDuration)
```

Although, I recall @cwize1 saying he's got a better idea to limit the network connections, so I'm guessing this comment might become completely obsolete soon...

Reply:
Also, since we're logging errors below as well, I'm curious if we can simplify and reduce the number of repeated error logs in case of multiple download failures:
```suggestion
	logger.Log.Debugf("Pre-caching '%s' from '%s'", fileName, url)
	err := retry.Run(func() error {
		// Limit the number of concurrent network operations to avoid overloading the network.
		netOpsSemaphore <- struct{}{}
		defer func() {
			<-netOpsSemaphore
		}()

		return network.DownloadFile(url, fullFilePath, nil, nil)
	}, downloadRetryAttempts, downloadRetryDuration)
```

Reply:
I think its actually beneficial to have the semaphore outside the retry loop. If its inside, we will just keep hammering the connection as quickly as possible for each worker while what we actually want is the system to globally slow down when errors start occurring.

I'm also personally interested in keeping the logs, we generally shouldn't be seeing these since we won't try to download anything that doesn't exist. So anything we see here is going to be unexpected and transient. I think it would be nice to see how common the errors are, and how long they persist.